Implement DOMDocument::adoptNode()

nielsdos opened the pull request:
For the past 20 years this threw a "not yet implemented" exception. But the function was actually there (albeit not documented) and could be called...

Girgias commented on ext/dom/document.c:
```diff
@@ -1049,9 +1049,10 @@ static void php_dom_transfer_document_ref(xmlNodePtr node, dom_object *dom_objec
 			ZEND_ASSERT(iteration_object != NULL);
-			php_libxml_decrement_doc_ref(iteration_object);
+			dom_object_document->document->refcount++;
+			 php_libxml_decrement_doc_ref(iteration_object);
```
Nit: indent
```suggestion
			php_libxml_decrement_doc_ref(iteration_object);
```

Girgias commented on ext/dom/document.c:
```diff
@@ -1049,9 +1049,10 @@ static void php_dom_transfer_document_ref(xmlNodePtr node, dom_object *dom_objec
+			 php_libxml_decrement_doc_ref(iteration_object);
 			iteration_object->document = dom_object_document->document;
-			php_libxml_increment_doc_ref(iteration_object, document);
+			// php_libxml_increment_doc_ref(iteration_object, document);
```
Left over comment?

Girgias commented on ext/dom/document.c:
```diff
@@ -1039,18 +1039,68 @@ PHP_METHOD(DOMDocument, getElementById)
+		|| nodep->type == XML_ENTITY_NODE
+		|| nodep->type == XML_NOTATION_NODE)) {
+		php_dom_throw_error(NOT_SUPPORTED_ERR, true);
+		return;
```
As this always seems to throw
```suggestion
		RETURN_THROWS();
```

Girgias reviewed (approved):
Looks good to me, the only question I have is about the "shadow root" bit in the spec, as it doesn't look to be handled. But this may be something that libxml doesn't support

nielsdos commented:
Note that I'm seeing CI failures here, which didnt reproduce on my local system, may be libxml version related. I started debugging this using the CI on my fork yesterday and will continue with this today. This is the reason for the indent and commented code, and hence why I converted this to draft.
I'll ping you when I resolved the issue. Thanks for review nevertheless.

Girgias commented:
Ah indeed, just been going through my emails and didn't check CI

kocsismate commented on ext/dom/php_dom.stub.php:
```diff
@@ -787,7 +787,7 @@ public function validate(): bool {}
 
-    /** @return DOMNode|false */
+    /** @return DOMNode */
     public function adoptNode(DOMNode $node) {}
```
please add a tentative return type here :)

nielsdos replied:
Thanks. I'll do that (if I get around to finishing this implementation :sweat_smile: )

nielsdos commented:
I fixed the remaining CI failures.
I have one question though: what am I supposed to do with these "tentative return types"? I'm not really sure when I'm supposed to use them.
In the previous implementation I modified the return type. But in the current version of this PR implementation the stub stayed the same. I suppose I don't need to make any changes to the stub?

Girgias commented:
> I fixed the remaining CI failures. I have one question though: what am I supposed to do with these "tentative return types"? I'm not really sure when I'm supposed to use them. In the previous implementation I modified the return type. But in the current version of this PR implementation the stub stayed the same as the original. I suppose I don't need to make any changes to the stub?

Tentative return types are a way to alleviate BC breaks for adding return types to non-final classes.

As return types are co-variant, adding one when there wasn't any means that every child class now needs to also add a return type, marking at tentative gives a warning that one needs to add one without actually preventing compilation.

Not sure why this method didn't have an actual return type, but I would do what @kocsismate asked and add one, as it also helps with syncing the docs more accurately.

nielsdos commented:
@Girgias Thanks, I get it now. I now added such a tentative return type.

kocsismate commented:
Most probably, the return type hasn't been declared until now because of the missing implementation :) Now, it's time indeed to warn users that there will be a native type added soon.
